### PR TITLE
ios-low-latency

### PR DESCRIPTION
**Pull request by ddonze**
The original iOS implementation of the plugin uses the AVPlayer class, which is excellent for media playback (audio AND video) and is particularly well-suited for streaming content.  However, due to its overhead, it is not the best solution for short local audio clips (such as sound effects in a video game). There is significant and noticeable lag when playing these kind of sounds.

This branch seeks to remedy that problem by using AVAudioPlayer instead of AVPlayer in situations where the audio is tagged as 'local' and 'low-latency.'  To do this, I've wrapped AVPlayer and AVAudioPlayer in a PlayerProtocol interface.

This branch is not complete. I created it specifically for a game project I am working on. Therefore, I have not tested the AVPlayer implementation very well and I know parts are incomplete. For example, none of the callbacks are wired up.

Please feel free to clean up and fix things as you please (or toss it all in the trash if you want :)

### CODE REVIEW

File: lib/audio_cache.dart [hunk 1]
```diff
@@ -117,14 +117,14 @@ class AudioCache {
...
     AudioPlayer player = _player(mode);
-    player.setReleaseMode(ReleaseMode.LOOP);
     player.play(
       file.path,
       isLocal: true,
       volume: volume,
       respectSilence: isNotification ?? respectSilence,
       stayAwake: stayAwake,
     );
+    player.setReleaseMode(ReleaseMode.LOOP);
```
Comment (erickzanardo) [inline lib/audio_cache.dart]: Any reason for this change?

**Comment by erickzanardo** [comment]
Looking good for me!

File: lib/audio_cache.dart (hunk 1 above)
Comment (ddonze) [reply lib/audio_cache.dart]: the iOS audio player is created inside the play() method. So you can't set the loop mode until AFTER you call play()

File: ios/Classes/AudioplayersPlugin.m [hunk 2]
```diff
@@ -366,12 +229,12 @@ -(void) onSoundComplete: (NSString *) playerId {
...
-  [ self seek:playerId time:CMTimeMakeWithSeconds(0,1) ];
+  //[ self pause:playerId ];
+  //[ self seek:playerId time:0 ];
 
-  if ([ playerInfo[@"looping"] boolValue]) {
-    [ self resume:playerId ];
-  }
+  //if ([ playerInfo[@"looping"] boolValue]) {
+  //  [ self resume:playerId ];
+  //}
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: commented out code?

File: ios/Classes/AudioplayersPlugin.m [hunk 3]
```diff
@@ -383,20 +246,20 @@ -(void)observeValueForKeyPath:(NSString *)keyPath
   if ([keyPath isEqualToString: @"player.currentItem.status"]) {
     NSString *playerId = (__bridge NSString*)context;
     NSMutableDictionary * playerInfo = players[playerId];
-    AVPlayer *player = playerInfo[@"player"];
+    AVAudioPlayer *player = playerInfo[@"player"];
 
-    NSLog(@"player status: %ld",(long)[[player currentItem] status ]);
+    //NSLog(@"player status: %ld",(long)[[player currentItem] status ]);
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: just remove it

File: example/ios/Flutter/flutter_export_environment.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+# This is a generated file; do not edit or check into version control.
+export "FLUTTER_ROOT=/Users/daviddonze/flutter"
+export "FLUTTER_APPLICATION_PATH=/Users/daviddonze/Documents/fluttertests/maze/audioplayers/example"
+export "FLUTTER_TARGET=/Users/daviddonze/Documents/fluttertests/maze/audioplayers/example/lib/main.dart"
+export "FLUTTER_BUILD_DIR=build"
+export "SYMROOT=${SOURCE_ROOT}/../build/ios"
+export "FLUTTER_FRAMEWORK_DIR=/Users/daviddonze/flutter/bin/cache/artifacts/engine/ios"
+export "TRACK_WIDGET_CREATION=true"
```
Comment (JesseScott) [inline example/ios/Flutter/flutter_export_environment.sh]: this shouldn't be committed....

File: example/ios/Podfile.lock [hunk 5]
```diff
@@ -6,23 +6,23 @@ PODS:
...
 SPEC CHECKSUMS:
   audioplayers: 84f968cea3f2deab00ec4f8ff53358b3c0b3992c
-  Flutter: 58dd7d1b27887414a370fcccb9e645c08ffd7a6a
+  Flutter: 0e3d915762c693b495b44d77113d4970485de6ec
   path_provider: f96fff6166a8867510d2c25fdcc346327cc4b259
 
 PODFILE CHECKSUM: 1003743f06ec24ad436e3271b83a8ca2bee2fed4
 
-COCOAPODS: 1.7.0.beta.2
+COCOAPODS: 1.7.4
```
Comment (JesseScott) [inline example/ios/Podfile.lock]: changes to this file shouldn't be committed

File: example/ios/Runner.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -279,7 +279,7 @@
 			);
 			inputPaths = (
 				"${PODS_ROOT}/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh",
-				"${PODS_ROOT}/../../../../../flutter/bin/cache/artifacts/engine/ios-release/Flutter.framework",
+				"${PODS_ROOT}/../../../../../../../flutter/bin/cache/artifacts/engine/ios/Flutter.framework",
```
Comment (JesseScott) [inline example/ios/Runner.xcodeproj/project.pbxproj]: changes to this file shouldn't be committed

File: ios/Classes/AudioplayersPlugin.m [hunk 7]
```diff
@@ -259,44 +190,22 @@ -(void) play: (NSString*) playerId
...
-  NSLog(@"ios -> updateDuration...%f", CMTimeGetSeconds(duration));
-  if(CMTimeGetSeconds(duration)>0){
+  NSTimeInterval duration = player.duration;
+  NSLog(@"ios -> updateDuration...%f", duration);
+  if(duration > 0){
     NSLog(@"ios -> invokechannel");
-   int mseconds= CMTimeGetSeconds(duration)*1000;
-    [_channel_audioplayer invokeMethod:@"audio.onDuration" arguments:@{@"playerId": playerId, @"value": @(mseconds)}];
+   //int mseconds= duration*1000;
+    //[_channel_audioplayer invokeMethod:@"audio.onDuration" arguments:@{@"playerId": playerId, @"value": @(mseconds)}];
```
Comment (JesseScott) [inline ios/Classes/AudioplayersPlugin.m]: remove it

File: ios/Classes/WrappedAVAudioPlayer.m [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                       successfully:(BOOL)flag
+                       {
+
+                       }
+- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player 
+                                 error:(NSError *)error;
+                                 {
+
+                                 }
+@end
```
Comment (JesseScott) [inline ios/Classes/WrappedAVAudioPlayer.m]: do we really need the `!!!` in the logs?

File: ios/Classes/WrappedAVPlayer.m [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    else {
+        _url = newUrl;
+        AVPlayerItem *playerItem;
+        //if (isLocal) {
+            playerItem = [[AVPlayerItem alloc] initWithURL:[NSURL fileURLWithPath:_url]];
+        //}
+        //else {
+        //    playerItem = [[AVPlayerItem alloc] initWithURL:[NSURL URLWithString:_url]];
+        //}
```
Comment (JesseScott) [inline ios/Classes/WrappedAVPlayer.m]: ^ remove commented out code

File: ios/Classes/WrappedAVPlayer.m [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
```
Comment (JesseScott) [inline ios/Classes/WrappedAVPlayer.m]: remove commented out code

**Comment by om-ha** [comment]
This is an important PR.

`AVAudioPlayer` and `AVPlayer` are used differently when it comes to audio playback.

See [this](https://stackoverflow.com/a/42829233/10830091) for comparison between the two.

The only downside of using `AVAudioPlayer` over `AVPlayer` in my opinion, is the requirement to use multiple `AVAudioPlayer` instances to play multiple files.

Both classes report incorrect `currentTime` after seeking. Each class has a fix for this different from the other.

The following needs to be considered and taken into account IMO before picking one:
- The Comparison above between the two illustrating the advantages/disadvantages. 
- The required bandaid fixes to get them working as required.

**Comment by jaymacdonald67** [comment]
Can I use this version? My app has short sound clips and there is a noticeable delay in iOS.

**Comment by ddonze** [comment]
Yes, please try it out. Be warned, however, that this is made specifically with game audio in mind and hasn't been tested with anything but local sound files. And certain features haven't been developed at all yet - such as seeking.
There is also a windows branch which has a basic plugin template set up for windows desktop, but nothing has really been implemented yet.

**Comment by jaymacdonald67** [comment]
It should work well for my app. It is a trivia app, all sound files are local and short. And no need for seeking or anything fancy.

**Comment by jaymacdonald67** [comment]
So, dumb question, how do I get this branch into my project?

**Comment by ddonze** [comment]
First, clone my repo into a folder on your computer:
`git clone https://github.com/ddonze/audioplayers.git`
Then checkout the ios-low-latency branch:
`git checkout ios-low-latency`
Then, modify your projects pubspec.yaml to point to the folder where you cloned my repo:
`dependencies:   
  audioplayers:   
    path: ../plugins/audioplayers`

**Comment by jaymacdonald67** [comment]
Is the "onPlayerCompletion" event not implemented? 

In my trivia app, when a sound is complete, I move to the next question.

 _advancedPlayer.onPlayerCompletion.listen((event)
      {
        _playerStopped();
      });

**Comment by ddonze** [comment]
Oh that's possible. If you can wait until tonight or tomorrow, I should be able to get that implemented for you.

**Comment by jaymacdonald67** [comment]
For sure. Thanks

**Comment by jaymacdonald67** [comment]
I am not an iOS programmer, but is this a typo in your _Audioplayersplugin.m_ ? 

Screen Shot 2020-03-02 at 5 38 10 PM

**Comment by ddonze** [comment]
Ha. That is a typo, but I didn't write that bit of code. I will fix it however because I can't stand typos.

**Comment by erickzanardo** [comment]
This PR is quite outdated and haven't got updates in a long time. Closing this. If the author wants to continue working on this, just lets us know and we can re open the PR.

**Pull request closed without merging**